cluster/v1alpha1/helpers.go

```diff
@@ -110,119 +116,182 @@ func (r *RolloutHandler) getRolloutAllClusters(rolloutStrategy RolloutStrategy,
-	rolloutResult := progressivePerGroup(clusterGroups, maxTimeDuration, statusFunc)
-	if len(rolloutResult.ClustersToRollout) > 0 {
-		return &strategy, rolloutResult, nil
-	}
+	// Perform progressive rollout per group for mandatory decision groups first
+	if len(clusterGroups) > 0 {
+		rolloutResult := progressivePerGroup(clusterGroups, failureTimeout, currentClusterStatus)
 
-	// Parse timeout for non-mandatory decision groups
-	failureTimeout, err := parseTimeout(strategy.ProgressivePerGroup.Timeout.Timeout)
-	if err != nil {
-		return &strategy, RolloutResult{}, err
+		if len(rolloutResult.ClustersToRollout) > 0 || len(rolloutResult.ClustersTimeOut) > 0 {
+			rolloutResult.ClustersRemoved = removedClusterStatus
+			return &strategy, rolloutResult, nil
+		}
 	}
 
-	// Upgrade the rest of the decision groups
-	restClusterGroups := r.pdTracker.ExistingClusterGroupsBesides(clusterGroups.GetOrderedGroupKeys()...)
+	// RollOut the rest of the decision groups
+	restClusterGroups := r.pdTracker.ExistingClusterGroupsBesides(groupKeys...)
 
 	// Perform progressive rollout per group for the remaining decision groups
-	rolloutResult = progressivePerGroup(restClusterGroups, failureTimeout, statusFunc)
+	rolloutResult := progressivePerGroup(restClusterGroups, failureTimeout, currentClusterStatus)
+	rolloutResult.ClustersRemoved = removedClusterStatus
+
 	return &strategy, rolloutResult, nil
 }
 
-func progressivePerCluster(clusterGroupsMap clusterv1beta1.ClusterGroupsMap, length int, timeout time.Duration, statusFunc ClusterRolloutStatusFunc) RolloutResult {
-	rolloutClusters := map[string]ClusterRolloutStatus{}
-	timeoutClusters := map[string]ClusterRolloutStatus{}
+func (r *RolloutHandler[T]) getRemovedClusters(clusterGroupsMap clusterv1beta1.ClusterGroupsMap, existingClusterStatus []ClusterRolloutStatus) ([]ClusterRolloutStatus, []ClusterRolloutStatus) {
+	currentClusterStatus := []ClusterRolloutStatus{}
+	removedClusterStatus := []ClusterRolloutStatus{}
+
+	clusters := clusterGroupsMap.GetClusters().UnsortedList()
+	for _, clusterStatus := range existingClusterStatus {
+		exist := false
+		for _, cluster := range clusters {
+			if clusterStatus.ClusterName == cluster {
+				exist = true
+				currentClusterStatus = append(currentClusterStatus, clusterStatus)
+				break
+			}
+		}
 
-	if length == 0 {
-		return RolloutResult{
-			ClustersToRollout: rolloutClusters,
-			ClustersTimeOut:   timeoutClusters,
+		if !exist {
+			removedClusterStatus = append(removedClusterStatus, clusterStatus)
+		}
+	}
+	return currentClusterStatus, removedClusterStatus
+}
+
+func progressivePerCluster(clusterGroupsMap clusterv1beta1.ClusterGroupsMap, length int, timeout time.Duration, existingClusterStatus []ClusterRolloutStatus) RolloutResult {
+	rolloutClusters := []ClusterRolloutStatus{}
+	timeoutClusters := []ClusterRolloutStatus{}
+	existingClusters := make(map[string]bool)
+
+	for _, status := range existingClusterStatus {
```

Comment:
Should here sort the existingClusterStatus by name by alphabetical order to ensure consistency?

Reply:
Could help explain a bit more for the existingClusterStatus, for example:
1. For a fresh install, should existingClusterStatus include any clusters? 
2. For a rollout case, if 100 clusters have old workload with status success, should they be put into existingClusterStatus with status Succeed or should they not be put into existingClusterStatus?